### PR TITLE
Add android test app to validate Java API for ORT-Mobile Android

### PR DESCRIPTION
**Pull request by YUNQIUGUO**
**Description**: Describe your changes.
Add android test application to validate Java API for ORT-Mobile Android

1. Add a simple test android app based on sigmoid.ort model. (instrumental android test, meant for running on android emulator).
2. Enable cmake complie and build apks required to build the android test app.
3. Modify build.py script to enable the test run in adb shell.

### CODE REVIEW

File: java/src/test/android/local.properties [hunk 1]
```diff
@@ -0,0 +1,10 @@
+## This file is automatically generated by Android Studio.
+# Do not modify this file -- YOUR CHANGES WILL BE ERASED!
+#
+# This file should *NOT* be checked into Version Control Systems,
+# as it contains information specific to your local configuration.
+#
```
Comment (skottmckay) [inline java/src/test/android/local.properties:6]: Looks like some stuff here isn't intended to be checked in. Is the '.gitignore' file in this directory meant to apply?

Also wondering if the files ending in '.lock' are meant to be checked in.

File: java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidater/SimpleTest.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+public class SimpleTest {
+
+    @Test
+    public void runSigmoidModelTest() throws OrtException, IOException {
+
+        try (OrtEnvironment env = OrtEnvironment.getEnvironment();
+             OrtSession.SessionOptions opts = new OrtSession.SessionOptions()) {
+
+            opts.setOptimizationLevel(OrtSession.SessionOptions.OptLevel.BASIC_OPT);
+
```
Comment (skottmckay) [inline java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidater/SimpleTest.java:29]: fwiw we don't run optimizations on an ORT format model, so this wouldn't affect anything.

File: java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidater/SimpleTest.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                String inputName = session.getInputNames().iterator().next();
+                float[][][] testdata = new float[3][4][5];
+                for (float[][] array2d : testdata) {
+                    for (float[] array : array2d) {
+                        Arrays.fill(array, (float) 0);
+                    }
+                }
+                try (OnnxTensor inputTensor =
+                             OnnxTensor.createTensor(env, testdata);
+                     OrtSession.Result output = session.run(Collections.singletonMap(inputName, inputTensor))) {
```
Comment (skottmckay) [inline java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidater/SimpleTest.java:40]: nit: is the line break necessary here?

File: java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidater/SimpleTest.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+
+    private byte[] readModel(String fileName) throws IOException {
+        AssetManager assetManager =
```
Comment (skottmckay) [inline java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidater/SimpleTest.java:58]: Do we need to go via this manual process or can we just create the session by passing in the filename (as ORT has code to handle reading bytes from a file)?

**Comment by skottmckay** [comment]
Can you add a readme.txt to java/src/test/android that provides an overview of what's in that directory and how it's expected to be used?

File: tools/ci_build/build.py [hunk 5]
```diff
@@ -1206,6 +1209,13 @@ def run_adb_shell(cmd):
                 os.path.join(source_dir, 'cmake', 'external', 'onnx', 'onnx', 'backend', 'test'),
                 device_dir, cwd=cwd)
             adb_push('onnxruntime_test_all', device_dir, cwd=cwd)
+            adb_install(
+                os.path.join(source_dir, 'build', 'Windows', "Debug", "java", "androidtest", "android", "app", "build",
```
Comment (edgchen1) [inline tools/ci_build/build.py:1213]: a more general way to specify `os.path.join(source_dir, 'build', 'Windows', 'Debug', ...)` would be `os.path.join(cwd, ...)`. as it is, it only works on Windows with a particular build directory.
I'd also say that `cwd` is not a great name in this context. maybe a better fix would be to pass in build_dir as a parameter to this function instead.

Comment (edgchen1) [reply tools/ci_build/build.py:1213]: also are some of the "debug"'s further along the path dependent on the build configuration?

Comment (YUNQIUGUO) [reply tools/ci_build/build.py]: > also are some of the "debug"'s further along the path dependent on the build configuration?

Yes, thanks for pointing out that. It is only for `build_java` . Added a check for that.

File: java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidater/SimpleTest.java (hunk 4 above)
Comment (YUNQIUGUO) [reply java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidater/SimpleTest.java:55]: I tried to directly pass a model path into `ort.createsession()`, seems not working. (I am just guessing) is it because we are loading an ort format model but not an onnx format model?

File: java/src/test/android/app/src/test/java/ai/onnxruntime/example/javavalidater/ExampleUnitTest.kt [hunk 6]
```diff
@@ -0,0 +1,17 @@
+package ai.onnxruntime.example.javavalidater
```
Comment (edgchen1) [inline java/src/test/android/app/src/test/java/ai/onnxruntime/example/javavalidater/ExampleUnitTest.kt:1]: do we need this example test?

File: java/src/test/android/gradlew [hunk 7]
```diff
@@ -0,0 +1,172 @@
+#!/usr/bin/env sh
```
Comment (edgchen1) [inline java/src/test/android/gradlew:1]: do we need to check in gradlew and gradlew.bat?
I noticed in the top level .gitignore file they are excluded from the java/ directory.
if we don't need them, perhaps we can also update that .gitignore file.

Comment (YUNQIUGUO) [reply java/src/test/android/gradlew:1]: No, we don't need gradlew and gradlew.bat. will update.

File: java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidater/SimpleTest.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
+package ai.onnxruntime.example.javavalidater;
```
Comment (edgchen1) [inline java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidater/SimpleTest.java:1]: nit: validater -> validator

File: java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidater/SimpleTest.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+import ai.onnxruntime.*;
```
Comment (edgchen1) [inline java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidater/SimpleTest.java:3]: if we import ai.onnxruntime.*, would that cover other imports like ai.onnxruntime.OnnxTensor?
that said, if it's not inconvenient it might be preferable to explicitly import the names we require instead of everything.

File: java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidater/SimpleTest.java (hunk 4 above)
Comment (edgchen1) [reply java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidater/SimpleTest.java:55]: it should support ORT format models too. what kind of error do you see?

File: java/src/test/android/README.md [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+
+## Background
+
+For general usage and build purpose of ORT-Mobile Android, please see the [documentation](https://www.onnxruntime.ai/docs/how-to/build.html#android) here.
+
+### Test Android Application Overview
+
+This android application is mainly aimed for testing:
+
+- Model used: A simple [sigmoid ONNX model](https://github.com/onnx/onnx-docker/blob/master/onnx-ecosystem/inference_demos/simple_onnxruntime_inference.ipynb) (converted to ORT format under `app\src\androidTest\assets` folder)
```
Comment (edgchen1) [inline java/src/test/android/README.md:13]: it might be good to link to the exact model file with the version. I guess it's this?
https://github.com/onnx/onnx/blob/f9b0cc99344869c246b8f4011b8586a39841284c/onnx/backend/test/data/node/test_sigmoid/model.onnx
also might be useful to provide the command line invocation for generating the ORT format model, in case we ever need to update it later.

File: tools/ci_build/build.py [hunk 11]
```diff
@@ -1210,6 +1213,12 @@ def run_adb_shell(cmd):
             adb_push('onnx_test_runner', device_dir, cwd=cwd)
             adb_shell('chmod +x {}/onnx_test_runner'.format(device_dir))
             run_adb_shell('{0}/onnxruntime_test_all'.format(device_dir))
+            if args.build_java:
+                adb_install(
+                    os.path.join(build_dir, "Debug", "java", "androidtest", "android", "app", "build", "outputs", "apk", "debug", "app-debug.apk"))
```
Comment (edgchen1) [inline tools/ci_build/build.py:1218]: we probably don't want to hardcode "Debug" here, or it won't work for other build configurations.
could use `get_config_build_dir(build_dir, config)` to get the start of the path.

File: java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidater/SimpleTest.java (hunk 9 above)
Comment (YUNQIUGUO) [reply java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidater/SimpleTest.java:3]: Yes, it will cover everything including ai.onnxruntime.OnnxTensor. but thanks for the suggestions, since we basically only need onnxtensor, session, etc. here, I will explicitly import them.

File: java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidater/SimpleTest.java (hunk 8 above)
Comment (edgchen1) [reply java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidater/SimpleTest.java]: just a few other places:
```
$ git grep -i -n javavalidater
java/src/test/android/app/build.gradle:11:        applicationId "ai.onnxruntime.example.javavalidater"
java/src/test/android/app/src/main/AndroidManifest.xml:3:    package="ai.onnxruntime.example.javavalidater">
java/src/test/android/app/src/main/AndroidManifest.xml:11:        android:theme="@style/Theme.JavaValidater">
java/src/test/android/app/src/main/res/values-night/themes.xml:3:    <style name="Theme.JavaValidater" parent="Theme.MaterialComponents.DayNight.DarkActionBar">
java/src/test/android/app/src/main/res/values/strings.xml:2:    <string name="app_name">JavaValidater</string>
java/src/test/android/app/src/main/res/values/themes.xml:3:    <style name="Theme.JavaValidater" parent="Theme.MaterialComponents.DayNight.DarkActionBar">
java/src/test/android/settings.gradle:2:rootProject.name = "JavaValidater"
tools/ci_build/build.py:1228:                    'am instrument -w ai.onnxruntime.example.javavalidater.test/androidx.test.runner.AndroidJUnitRunner'
```

File: java/src/test/android/README.md [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+Use the android's [build instructions](https://www.onnxruntime.ai/docs/how-to/build.html#android-build-instructions) with `--build_java` and `--android_run_emulator` option.
+
+Please note that you may need to set the `--android_abi=x86_64` (the default option is `arm64-v8a`). This is because android instrumentation test is run on an android emulator which requires an abi of `x86_64`.
+
+#### Build Output
+
+The build will generate two apks which is required to run the test application in `$YOUR_BUILD_DIR/java/androidtest/android/app/build/outputs/apk`:
+
+* `androidtest/debug/app-debug-andrdoidtest.apk`
```
Comment (edgchen1) [inline java/src/test/android/README.md:34]: "app-debug-andrdoidtest.apk" -> "app-debug-androidtest.apk"

Comment (edgchen1) [inline java/src/test/android/README.md:34]: just to confirm, is the "debug" part of these names there regardless of the ORT build configuration?

Comment (YUNQIUGUO) [reply java/src/test/android/README.md:34]: > just to confirm, is the "debug" part of these names there regardless of the ORT build configuration?

Yes, it is not related to the ORT build configuration. when we use gradle to build apks for android application, it will always generate a debug apk.

Comment (YUNQIUGUO) [reply java/src/test/android/README.md]: https://github.com/microsoft/onnxruntime/pull/7477/files#diff-0f2b7c808c4ee2b3f8135f74bcd5c4fb73bef5dbc13fb5f82014e5ff99b61b61R219 Here, we use `gradle assembleDebug assembleDebugAndroidTest`, and the "debug" comes from this.

File: java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidater/SimpleTest.java (hunk 4 above)
Comment (YUNQIUGUO) [reply java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidater/SimpleTest.java]: > it should support ORT format models too. what kind of error do you see?

Sorry just saw this. The error message is `ai.onnxruntime.OrtException: Error code - ORT_INVALID_ARGUMENT - message: Invalid fd was supplied: -1` and I passed in a model path like `..\..\assets\sigmoid.ort`

Comment (edgchen1) [reply java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidater/SimpleTest.java]: I see - perhaps the path is not valid. looking around online a bit, it seems that AssetManager is the way to access assets. looks like they are stored in some other format (https://gamedev.stackexchange.com/questions/165071/how-to-get-asset-folder-path-in-android-native-activity)

File: java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidater/SimpleTest.java [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+            Log.e("Test", "Cannot load model from assets");
+        }
+        ByteArrayOutputStream buffer = new ByteArrayOutputStream();
+        int nRead;
+        byte[] data = new byte[16384];
+        while ((nRead = inputStream.read(data, 0, data.length)) != -1) {
+            buffer.write(data, 0, nRead);
+        }
+        return buffer.toByteArray();
+    }
```
Comment (gwang-msft) [inline java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidater/SimpleTest.java]: If you use kotlin, this can be as simple as 
InstrumentationRegistry.getInstrumentation().context.assets.open(fileName).readBytes()

File: java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidater/SimpleTest.java [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+            opts.setOptimizationLevel(OrtSession.SessionOptions.OptLevel.BASIC_OPT);
+
+            try (OrtSession session = env.createSession(readModel("sigmoid.ort"), opts)) {
+                String inputName = session.getInputNames().iterator().next();
+                float[][][] testdata = new float[3][4][5];
+                for (float[][] array2d : testdata) {
+                    for (float[] array : array2d) {
+                        Arrays.fill(array, (float) 0);
+                    }
+                }
```
Comment (gwang-msft) [inline java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidater/SimpleTest.java]: Do you want to create the input array with different values (not just all 0s, for example 0 to 3*4*5 or even better with some negative input) and calculate the expected result here directly and use it to compare

File: java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidater/SimpleTest.java (hunk 4 above)
Comment (gwang-msft) [reply java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidater/SimpleTest.java]: The only path you can have read access is from '/data/local/tmp' (we use that for our UT), which is not recommended for actual Android App.

File: tools/ci_build/github/linux/ort_minimal/build_minimal_ort_android_baseline_and_report_bin_size.sh [hunk 15]
```diff
@@ -54,3 +54,6 @@ fi
 
 # Clear the build
 rm -rf /build/MinSizeRel
+
+# Use a newer version of gradle for Android Test
+/bin/bash /onnxruntime_src/tools/ci_build/github/android/setup_gradle_wrapper.sh /onnxruntime_src
```
Comment (gwang-msft) [inline tools/ci_build/github/linux/ort_minimal/build_minimal_ort_android_baseline_and_report_bin_size.sh:59]: You will need to move this before the build starts

File: java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidator/SimpleTest.kt [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+                    }
+                }
+            }
+            //expected sigmoid output is y = 1.0 / (1.0 + exp(-x))
+            val expected =  Array(3) { Array(4) { FloatArray(5) } }
+            for (i in 0..2) {
+                for (j in 0..3) {
+                    for (k in 0..4) {
+                        expected[i][j][k] = (1.0 / (1.0 + kotlin.math.exp(-testdata[i][j][k]))).toFloat()
+                    }
```
Comment (gwang-msft) [inline java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidator/SimpleTest.kt:39]: Move this into the above loop

File: java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidator/SimpleTest.kt [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+            val inputTensor = OnnxTensor.createTensor(env, testdata)
+            inputTensor.use {
+                val output = session.run(Collections.singletonMap(inputName, inputTensor))
+                val rawOutput = output[0].value as Array<Array<FloatArray>>
+                for (i in 0..2) {
+                    for (j in 0..3) {
+                        for (k in 0..4) {
+                            val actual = (kotlin.math.round(rawOutput[i][j][k] * 1000.0) / 1000.0).toFloat()
+                            val expected = (kotlin.math.round(expected[i][j][k] * 1000.0) / 1000.0).toFloat()
+                            Assert.assertEquals(actual, expected)
```
Comment (gwang-msft) [inline java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidator/SimpleTest.kt:45]: > Assert.assertEquals [](http://example.com/codeflow?start=27&length=20)

You can still use delta with Assert.assertEquals

File: java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidator/SimpleTest.kt [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+
+@RunWith(AndroidJUnit4::class)
+class SimpleTest {
+    @Test
+    @Throws(OrtException::class, IOException::class)
+    fun runSigmoidModelTest() {
+        val env = OrtEnvironment.getEnvironment()
+        env.use {
+            val opts = SessionOptions()
+            val session = env.createSession(readModel("sigmoid.ort"), opts)
```
Comment (gwang-msft) [inline java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidator/SimpleTest.kt:23]: opts, session need .use{...}

File: java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidator/SimpleTest.kt [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+                for (j in 0..3) {
+                    for (k in 0..4) {
+                        expected[i][j][k] = (1.0 / (1.0 + kotlin.math.exp(-testdata[i][j][k]))).toFloat()
+                    }
+                }
+            }
+
+            val inputTensor = OnnxTensor.createTensor(env, testdata)
+            inputTensor.use {
+                val output = session.run(Collections.singletonMap(inputName, inputTensor))
```
Comment (gwang-msft) [inline java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidator/SimpleTest.kt:38]: >output [](http://example.com/codeflow?start=20&length=6)

output need .use{...}

File: java/src/test/android/build.gradle [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+def cmakeJavaDir = "${cmakeBuildDir}/java"
+def cmakeAndroidTestDir = "${cmakeJavaDir}/androidtest"
+def cmakeBuildOutputDir = "${cmakeAndroidTestDir}/android/app/build"
+
+if (cmakeBuildDir != null) {
+    task cmakeBuild(type: Copy) {
+        from rootProject.buildDir
+        into cmakeBuildOutputDir
+    }
+
```
Comment (gwang-msft) [inline java/src/test/android/build.gradle:35]: Do you need this?

File: java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidator/SimpleTest.kt [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+                    val inputTensor = OnnxTensor.createTensor(env, testdata)
+                    inputTensor.use {
+                        val output = session.run(Collections.singletonMap(inputName, inputTensor))
+                        output.use {
+                            val rawOutput = output[0].value as Array<Array<FloatArray>>
+                            for (i in 0..2) {
+                                for (j in 0..3) {
+                                    for (k in 0..4) {
+                                        val actual = (kotlin.math.round(rawOutput[i][j][k] * 1000.0) / 1000.0).toFloat()
+                                        val expected = (kotlin.math.round(expected[i][j][k] * 1000.0) / 1000.0).toFloat()
```
Comment (gwang-msft) [inline java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidator/SimpleTest.kt:47]: Remove this?

File: java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidator/SimpleTest.kt [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+    fun runSigmoidModelTest() {
+        val env = OrtEnvironment.getEnvironment()
+        env.use {
+            val opts = SessionOptions()
+            opts.use {
+                val session = env.createSession(readModel("sigmoid.ort"), opts)
+                session.use {
+                    val inputName = session.inputNames.iterator().next()
+                    val testdata = Array(3) { Array(4) { FloatArray(5) } }
+                    val expected =  Array(3) { Array(4) { FloatArray(5) } }
```
Comment (gwang-msft) [inline java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidator/SimpleTest.kt:28]: >  Array(3) [](http://example.com/codeflow?start=34&length=10)

nit, extra space here,
you may try to run the "Reformat code" on this file

File: java/src/test/android/build.gradle [hunk 23]
```diff
@@ -0,0 +1,27 @@
...
+buildscript {
+	ext.kotlin_version = "1.4.32"
+
+	repositories {
+		google()
+		jcenter()
+	}
+	dependencies {
+		classpath "com.android.tools.build:gradle:4.1.3"
+		classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
```
Comment (gwang-msft) [inline java/src/test/android/build.gradle:11]: You can choose to use 4.0.0 for com.android.tools.build:gradle such that it does not require gradle 6.5, such that you can revert the change to tools/ci_build/github/linux/ort_minimal/build_minimal_ort_and_run_tests.sh

File: java/src/test/android/README.md [hunk 24]
```diff
@@ -0,0 +1,37 @@
...
+
+For general usage and build purpose of ORT-Mobile Android, please see the [documentation](https://www.onnxruntime.ai/docs/how-to/build.html#android) here.
+
+### Test Android Application Overview
+
+This android application is mainly aimed for testing:
+
+- Model used: A simple [sigmoid ONNX model](https://github.com/onnx/onnx/blob/f9b0cc99344869c246b8f4011b8586a39841284c/onnx/backend/test/data/node/test_sigmoid/model.onnx) (converted to ORT format under `app\src\androidTest\assets` folder).
+    - Here's a [documentation](https://github.com/microsoft/onnxruntime/blob/master/docs/ONNX_Runtime_for_Mobile_Platforms.md#1-create-ort-format-model-and-configuration-file-with-required-operators) about how you can convert an ONNX model into ORT format.
+- Main test file: An android instrumentation test under `app\src\androidtest\java\ai.onnxruntime.example.javavalidator\SimpleTest.java`
```
Comment (gwang-msft) [inline java/src/test/android/README.md:15]: >\SimpleTest.java` [](http://example.com/codeflow?start=118&length=17)

SimpleTest.kt

File: java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidator/SimpleTest.kt [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+                        output.use {
+                            val rawOutput = output[0].value as Array<Array<FloatArray>>
+                            for (i in 0..2) {
+                                for (j in 0..3) {
+                                    for (k in 0..4) {
+                                        val actual =
+                                            (kotlin.math.round(rawOutput[i][j][k] * 1000.0) / 1000.0).toFloat()
+                                        val expect =
+                                            (kotlin.math.round(expected[i][j][k] * 1000.0) / 1000.0).toFloat()
+                                        Assert.assertEquals(actual, expect, 1e-15.toFloat())
```
Comment (gwang-msft) [inline java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidator/SimpleTest.kt:51]:
```suggestion
                                        Assert.assertEquals(rawOutput[i][j][k], expected[i][j][k], 1e-15.toFloat())
```

Comment (YUNQIUGUO) [reply java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidator/SimpleTest.kt:51]: If I remove the round, then the assert will fail. For instance, it will generate error like expected:<0.0066928566> but was:<0.006692851>. Not sure if it is because of the data type difference or anything else...

File: java/src/test/android/build.gradle (hunk 23 above)
Comment (YUNQIUGUO) [reply java/src/test/android/build.gradle:11]: Also, if I change the gradle down to 4.0.0, then it seems like the NDK version will have a dismatch. It seems like 4.0.0 requires NDK version 21.0.6113669. Error message: `No version of NDK matched the requested version 21.0.6113669. Versions available locally: 21.4.7075529, 22.1.7171670` Any other idea how to fix this?

File: java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidator/SimpleTest.kt (hunk 25 above)
Comment (gwang-msft) [reply java/src/test/android/app/src/androidTest/java/ai/onnxruntime/example/javavalidator/SimpleTest.kt:51]: You can try a bigger tolerance like 1e-6 in ` Assert.assertEquals`

File: java/src/test/android/build.gradle (hunk 23 above)
Comment (gwang-msft) [reply java/src/test/android/build.gradle:11]: Unfortunately this is an Android Gradle plugin bug started from 3.6 and was fixed in 4.1 https://issuetracker.google.com/issues/144111441

You may either downgrade the com.android.tools.build:gradle to 3.5.3, or revert to previous behavior

**Review by gwang-msft — APPROVED** [review]
:shipit:

**Review by skottmckay — APPROVED** [review]
:shipit:
